lib/chef/provider/package/msu.rb

```diff
@@ -0,0 +1,146 @@
+#
+# Author:: Nimisha Sharad (<nimisha.sharad@msystechnologies.com>)
+# Copyright:: Copyright 2015-2016, Chef Software, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# msu_package leverages cab_package
+# The contents of msu file are extracted, which contains one or more cab files.
+# The extracted cab files are installed using Chef::Resource::Package::CabPackage
+# Reference: https://support.microsoft.com/en-in/kb/934307
+require "chef/provider/package"
+require "chef/resource/msu_package"
+require "chef/mixin/shell_out"
+require "chef/provider/package/cab"
+require "chef/util/path_helper"
+require "chef/mixin/uris"
+require "chef/mixin/checksum"
+
+class Chef
+  class Provider
+    class Package
+      class Msu < Chef::Provider::Package
+        use_inline_resources
+        include Chef::Mixin::ShellOut
+        include Chef::Mixin::Uris
+        include Chef::Mixin::Checksum
+
+        provides :msu_package, os: "windows"
+
+        def load_current_resource
+          @current_resource = Chef::Resource::MsuPackage.new(new_resource.name)
+
+          # download file if source is a url
+          msu_file = uri_scheme?(new_resource.source) ? download_source_file : @new_resource.source
+
+          # temp directory where the contents of msu file get extracted
+          @temp_directory = Dir.mktmpdir("chef")
+          extract_msu_contents(msu_file, @temp_directory)
+          @cab_files = read_cab_files_from_xml(@temp_directory)
+
+          unless @cab_files.empty?
+            # set the package version
+            # Here is a catch: How to set package version in case of multiple cab files?
+            cab_resource = @new_resource
+            cab_resource.source = @cab_files.first
+            cab_pkg = Chef::Provider::Package::Cab.new(cab_resource, nil)
+
+            new_resource.version(cab_pkg.package_version)
```

Comment:
To start with, you want to remove this line, the `new_resource.version(cab_pkg.package_version)`.

Then I'd look at something like `Chef::Provider::Package::Apt#get_current_versions` To answer the question of setting package version for multiple cab files, you could imitate something like the multi-package code in other providers.

In any case, all the code that opens up the cab files should be factored out to another method. You want a method that returns your `cab_pkg` , maybe `get_cab_package` and then in `load_current_resource` you set `current_resource.version(get_cab_package.installed_version)`.

Then you want to create a `candidate_version` method. this should return `get_cab_package.package_version`. `Chef::Provider::Package#target_version_array` when action is install, will set us up to try to install `new_version` if it is specified, that is if the user specific a version on the resource and it is not the current version, or if `new_version` was not set, it will try to install the `candidate_version`. The existing logic in `Chef::Provider::Package` should do all the right things for other situations like if you specify a version that differs than what's in the cab file, e.g. no matching candidates.